File: ex1.py
import sys

same_average = False
cnt = 0
k = sys.argv[1]

def argu_check(k,max_iter=200):
    if('.' in k):
        print("invalid input")
        assert()
    
    return int(max_iter)

try:
    max_iter = sys.argv[2]
    max_iter = argu_check(k,max_iter)
except IndexError:
    max_iter = argu_check(k)

k = int(sys.argv[1])

my_file = input()
clusters = [[] for i in range(k)]
new_file = open(my_file)

for line in new_file:
    if(cnt == k):
        break

    clusters[cnt].append([float(xi) for xi in line.split(",")]) 
    clusters[cnt].append(0)
    clusters[cnt].append([0.0 for xi in line.split(",")])
    cnt += 1

cnt = 0

def Euclidian_Distance(dot1, dot2):
    sum = 0.0
    for xi in range(len(dot1)):
        sum += pow((dot1[xi]-dot2[xi]),2)
    return sum

def Update_Mean(clusters, same_average):
    for i in range(k):
        for j in range(len(clusters[i][2])):
            if(clusters[i][2][j]/clusters[i][1] != clusters[i][0][j]):
                same_average = False
                clusters[i][0][j] = clusters[i][2][j]/clusters[i][1]
    return same_average


def Update_Sum_Of_Elements_In_Cluster(float_vector, loc, clusters):
    for i in range(len(float_vector)):
        clusters[loc][2][i] += float_vector[i]

def Finding_Cluster(float_vector, clusters):
    min_distance, num_of_cluster = -1, -1
    for i in range(k):
        distance = Euclidian_Distance(float_vector, clusters[i][0])
        if((distance < min_distance) or (min_distance < 0)): #distance is smaller or we are doing the first calc of distance
            min_distance = distance
            num_of_cluster = i
    
    clusters[num_of_cluster][1] += 1 #update number of elements in cluster
    Update_Sum_Of_Elements_In_Cluster(float_vector, num_of_cluster, clusters)


while (cnt < max_iter) and (not same_average):
    same_average = True #will be changed to False if one of the means change
    new_file.seek(0)
    for vector in new_file:
        float_vector = [float(xi) for xi in vector.split(",")]
        Finding_Cluster(float_vector, clusters)
    same_average = Update_Mean(clusters, same_average) #check if there's no change in mu
    if(same_average == True):
        break
    
    for cluster in clusters:
        cluster[1] = 0
        cluster[2] = [0 for x in cluster[2]] #initializing clusters properties before next iteration
    
    cnt += 1

for cluster in clusters:
    print(*["{:.4f}".format(num) for num in cluster[0]],sep=",")

File: ex1.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>

typedef struct 
{
    int num_of_points;
    double* sum_of_points;
    double* centroid;
}Cluster;


double Euclidian_Distance(double* vector1, double* vector2, int dimension);
void finding_cluster(double* vector, Cluster* clusters, int k, int dimension);
int update_mean(Cluster* clusters, int same_average, int k, int dimension);
void update_sum_of_elements_in_cluster(double* vector, int loc, Cluster* clusters, int dimension);
void free_memory(Cluster* clusters, double** data_points, int k, int num_of_points);

void main(int argc, char *argv[]){
    int max_iter = 0;
    int k = 0;
    Cluster* clusters;
    double** data_points;
    int num_of_points = 0;
    int dimension = 1;
    int same_average = 0;
    FILE *fp;
    char vector[1000];
    char *token;
    int cnt = 0;

    if(argc == 1){
        assert("Missing arguments");
    }

    if(atoi(argv[1]) < 0 || strchr(argv[1], '.') != NULL){
        printf("Invalid Input");
        assert(NULL);
    }
    k = atoi(argv[1]);
    printf("k: %d\n", k);

    if(argc == 2){ 
        max_iter = 200;
    }else if(atoi(argv[2]) < 0 || strchr(argv[2], '.') != NULL){
        printf("Invalid Input");
        assert(NULL);
    }else{
        max_iter = atoi(argv[2]);
    }
    
    printf("max iter: %d\n", max_iter);


    /* opening file for reading */
    scanf("%s", vector);
    fp = fopen(vector, "r");
    while (!feof(fp))
    {
        fgets(vector, 1000, fp);
        if(vector == NULL){
            /*line is empty*/
            break;
        }
        if(num_of_points == 0){
            for(int i = 0; vector[i] != '\n'; ++i){
                 if(vector[i] == ','){
                     dimension++;
                     }
                }
        }
        num_of_points += 1;
    }
    printf("num: %d\n", num_of_points);
    printf("dim: %d\n", dimension);

    data_points = calloc(num_of_points,sizeof(double*));
    for(int i = 0; i < num_of_points; i++){
        data_points[i] = calloc(dimension,sizeof(double));
    }

    rewind(fp);
    while (!feof(fp)){
        fgets(vector, 1000, fp);
        token = strtok(vector, ",");
        
        for(int i = 0; i < dimension; i++){
            if(token != NULL){
                data_points[cnt][i] = atof(token);
                token = strtok(NULL, ",");
            }
        }
        cnt++;
    }

    fclose(fp);

    /*
    Initializing k clusters
    */
    clusters = (Cluster*)calloc(k,sizeof(Cluster));
    for(int i = 0; i < k; i++){
        clusters[i].centroid = (double*)calloc(dimension, sizeof(double));
        assert(clusters[i].centroid != NULL);
        clusters[i].centroid = data_points[i]; //will be equal to the i'th vector
        clusters[i].num_of_points = 0;
        clusters[i].sum_of_points = (double*)calloc(dimension, sizeof(double));
        assert(clusters[i].sum_of_points != NULL);
        clusters[i].sum_of_points[i] = 0;
    }

    cnt = 0;
    while ((cnt < max_iter) && (!same_average))
    {
        same_average = 1;
        for(int i = 0; i < num_of_points; i++){
            finding_cluster(data_points[i], clusters, k, dimension);
        }
        
        same_average = update_mean(clusters, same_average, k, dimension);
        if(same_average == 1){
            break;
        }

        for(int i = 0; i < k; i++){
            clusters[i].num_of_points = 0;
            for(int j = 0; j < dimension; j++){
                clusters[i].sum_of_points[j] = 0;
            }
        }
        cnt++;
    }
    for(int i = 0; i < k; i++){
        for(int j = 0; j < dimension; j++){
            if(j == dimension-1){
                printf("%.4lf", clusters[i].centroid[j]);
            }else{
                printf("%.4lf,", clusters[i].centroid[j]);
            }
        }
        printf("\n");
    }
    free_memory(clusters, data_points, k, num_of_points);
}


void finding_cluster(double* vector, Cluster* clusters, int k, int dimension){
    double min_distance = -1.0;
    int num_of_cluster = -1;
    double distance;

    for(int i = 0; i < k; i++){
        distance = Euclidian_Distance(vector, clusters[i].centroid, dimension);
        if((distance < min_distance) || (min_distance < 0)){
            min_distance = distance;
            num_of_cluster = i;
        }
    }
    clusters[num_of_cluster].num_of_points++;
    update_sum_of_elements_in_cluster(vector, num_of_cluster, clusters, dimension);
}

void update_sum_of_elements_in_cluster(double* vector, int loc, Cluster* clusters, int dimension){
    for(int i = 0; i < dimension; i++){
        clusters[loc].sum_of_points[i] += vector[i];
    }
}


int update_mean(Cluster* clusters, int same_average, int k, int dimension){
    for(int i = 0; i < k; i++){
        for(int j = 0; j < dimension; j++){
            if((clusters[i].sum_of_points[j]/clusters[i].num_of_points)!=
                clusters[i].centroid[j]){
                    same_average = 0;
                    clusters[i].centroid[j] = clusters[i].sum_of_points[j]/clusters[i].num_of_points;
                }
        }
    }
    return same_average;
}

double Euclidian_Distance(double* vector1, double* centroid, int dimension){
    double sum = 0.0;
    for(int xi = 0; xi < dimension; xi++){
        sum += (vector1[xi]-centroid[xi])*(vector1[xi]-centroid[xi]);
    } 
    return sum;
}

void free_memory(Cluster* clusters, double** data_points, int k, int num_of_points){
    for(int i = 0; i < num_of_points; i++){
        free(data_points[i]);
    }
    free(data_points);

    for(int j = 0; j < k; j++){
        free(clusters[j].centroid);
        free(clusters[j].sum_of_points);
    }
    free(clusters);
}